Allow stepping through disassembly

danthe1st opened the pull request:
## What it does
This PR allows stepping through the bytecode of disassembled classes in the class file editor.

~~This is just an unfinished prototype but I wanted to create a (draft) PR early to be able to get feedback and show the basic approach.~~

This depends on another PR in jdt.ui (https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2708)

<img width="2160" height="1440" alt="image" src="https://github.com/user-attachments/assets/1b7c98b1-34f8-417f-8495-0ca112bf8fb1" />
<img width="2160" height="1440" alt="image" src="https://github.com/user-attachments/assets/38b942c4-c387-4d8d-be33-525f03c1e4a1" />


@SougandhS Since you have worked on similar things recently, I think you might be interested in that.

## How to test

- Start Eclipse with this change as well as [this patch](https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2708)
- Step through some code that steps into libraries without a source attachment
- It automatically jumps to the relevant bytecode instruction(s) corresponding to the stack frame
- Stepping through normal (non-library) classes and library classes with source attachments should still work

## Author checklist

- [x] I have thoroughly tested my changes
- [X] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [X] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

danthe1st committed:
Allow stepping through disassembly

This change adjusts JavaStackFrameEditorPresenter for ClassFileEditor,
so that bytecode instructions are highlighted during debug stepping.

SougandhS reviewed (commented):
<img width="791" height="163" alt="Image" src="https://github.com/user-attachments/assets/fd15c3ec-6604-4664-b37a-1072f9bed7f9" />

These load instructions will be skipped (wont be suspended) when `IJDIPreferencesConstants.PREF_STATEMENT_LEVEL_STEPPING` is enabled. - Can you confirm it ?

SougandhS commented:
>if that should be fixed, I guess this would make sense as a distinct change to keep the PRs small)

OK for me 👍 

>Other than that, it can step over individual bytecode instructions if no debugging information (like line numbers) is present:

Sounds nice

>If wanted, I can also try to add some tests for things like stepping without debugging information.

That would be good. please add few more tests if possible

danthe1st commented:
The tests failed with Jenkins because `AutomatedSuite` didn't run them in the UI thread. For now, I created a new `TestSuite` for this. What is the recommended approach for tests that should run fully in the UI thread?
Should I also move the test class to a different package?

trancexpress commented:
Odd, this seems to want the UI thread:

```
➜  org.eclipse.jdt.debug.tests git:(master) git grep useUIThread
pom.xml:          <useUIThread>true</useUIThread>
```

The shared launch also specifies running in the UI thread.

What is the error message from the test?

danthe1st commented:
> What is the error message from the test?

```
org.eclipse.ui.PartInitException: No active workbench page
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:388)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInSpecificEditor(EditorUtility.java:191)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:169)
	at org.eclipse.jdt.internal.ui.javaeditor.EditorUtility.openInEditor(EditorUtility.java:154)
	at org.eclipse.jdt.debug.tests.sourcelookup.ClassFileEditorHighlightingTest.testConstructorInPackage(ClassFileEditorHighlightingTest.java:99)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at org.eclipse.jdt.debug.tests.AbstractDebugTest.runBare(AbstractDebugTest.java:2993)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at org.eclipse.jdt.debug.tests.DebugSuite$1.run(DebugSuite.java:61)
	at java.base/java.lang.Thread.run(Thread.java:1516)
```
See https://github.com/eclipse-jdt/eclipse.jdt.debug/runs/70316773248

`AutomatedSuite` extends `DebugSuite` which runs all tests in a non-UI thread.
```java
/**
 * Debug test suite framework that runs test in a non UI thread.
 */
public abstract class DebugSuite extends TestSuite {
```

I could also move the test to a different test suite but I don't know which one would be better.

trancexpress commented:
I'm not sure, maybe the non-UI thread choice was made due t o `EventWaiter` not processing UI events while waiting... `DebugEventWaiter` on the other hand does process UI events...

From my POV no need to overcomplicate things for one test case, what is in the PR right now seems OK. If we find that the UI thread test suite is growing, we can have two top level test suites, where the UI one doesn't run tests in a separate thread.

danthe1st commented:
I think there's also `EvalTestSuite` which apparently does some UI tests but I don't know about that one and running those locally didn't seem to work properly for me (and I'm not sure whether that runs the tests fully in the UI thread or just does some additional processing for handling UI events).

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaStackFrameSourceDisplayAdapter.java:
```diff
@@ -0,0 +1,159 @@
+import org.eclipse.ui.texteditor.ITextEditor;
+
+/**
+ * @since 3.2
```
Should be possible to omit `@since`, since this is an internal package.

danthe1st replied:
I kept that when moving it from `LambdaStackFrameSourceDisplayAdapter`. Should I still remove it?

trancexpress replied:
Yes please, only needed for API.

danthe1st replied:
Done in 54fa9161eff690f018105650166fd236a455db4c.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/LambdaStackFrameSourceDisplayAdapter.java:
```diff
@@ -1,73 +0,0 @@
-/*******************************************************************************
```
@SougandhS are there any tests for this adapter? If not, how to test that nothing breaks when replacing it?

danthe1st replied:
I (more or less) renamed/generalized this class to `JavaStackFrameSourceDisplayAdapter` and added the logic for `ClassFileEditor`.

SougandhS replied:
@trancexpress Tests weren't added, it simply highlights the inline lambda exp on lambda breakpoint suspension 

see https://eclipse.dev/eclipse/markdown/?f=news/4.39/jdt.md#improved-highlighting-for-inline-chained-lambdas

```
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class BreakpointsTest {
	
	public static void main(String[] args) {
		callMe();
	}

	private static void callMe() { 
		BreakpointsTest t = new BreakpointsTest();
		t.lotOfLambdas();
	}
	
	void lotOfLambdas() {
		Optional.of("aa").map(this::a6).filter(x -> x.isBlank()).map(this::a6);

		List<String> list = Arrays.asList("a", "b", "c");
                // create a bp for any chained exp and see highlighting on suspension 
		list.stream().filter(s -> s.length() > 0).map(s -> s.toLowerCase()).filter(x -> x.charAt(0) > 1).map(s -> s.toUpperCase()).filter(k -> k.indexOf(0) < 0).map(s -> s.toLowerCase()).filter(p -> !p.isBlank()).forEach(System.out::println);
		
	}
	
	String a6(String s) {
		return s;
	}
}

```

SougandhS replied:
I had my testing with this patch, it was working fine 👍

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaDebugShowInAdapterFactory.java:
```diff
@@ -54,13 +54,8 @@ public <T> T getAdapter(Object adaptableObject, Class<T> adapterType) {
-			} catch (DebugException e) {
-				DebugUIPlugin.log(e);
+			if (adaptableObject instanceof JDIStackFrame) {
+				return (T) javaStackFrameSourceDisplayAdapter;// ensure the same instance is reused
```
Please add to the comment why it should be the same instance.

danthe1st replied:
This is the case because `T` is `ISourceDisplay` (checked by the `if`) which is the type of `javaStackFrameSourceDisplayAdapter` (I now changed that field to use the interface as well). I hope this is clear now.

trancexpress replied:
Sorry but I still don't understand the comment with the re-use. Is it for efficiency, not creating the object many times? Or there is some state that is important?

danthe1st replied:
This is because it stores state, specifically information about/for the listener ensuring that the highlighting is cleaned up when the instruction pointer is no longer at the class file.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaStackFrameSourceDisplayAdapter.java:
```diff
@@ -0,0 +1,159 @@
+					currentClassFileEditor.unhighlight();
+					currentClassFileEditor = null;
+					currentClassFileFrame = null;
+					service.removeDebugContextListener(this);
```
Can you add a test that debugs, then terminates the launch, then checks that we didn't leak a listener here?

danthe1st replied:
Is there a good way to check that the listener has been removed or should I just add a getter method for `classFileUnhighlighterRegistered` that's just there for the test?

trancexpress replied:
Unfortunately there is no access to `org.eclipse.debug.internal.ui.contexts.DebugWindowContextService.fListenersByPartId`... A method just for the test probably won't be enough, since the field can be overwritten and so the implementation can leak a listener without the test noticing.

Lets skip that.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaStackFrameSourceDisplayAdapter.java:
```diff
@@ -0,0 +1,159 @@
+	private boolean openClassFileAndHighlightStackFrame(JDIStackFrame jdiFrame, IClassFileEditorInput input) throws JavaModelException, PartInitException, DebugException {
+		IJavaElement je = JavaUI.getEditorInputJavaElement(input);
+		if (je != null) {
+			IEditorPart part = JavaUI.openInEditor(je);
```
Can you check what happens if this preference is off?

https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Freference%2Fpreferences%2Frun-debug%2Fref-run_debug.htm

```
 Reuse editor when displaying source code

The debugger displays source code in an editor when stepping through an application. When this option is on, the debugger will reuse the editor that it opened to display source from different source files. This prevents the debugger from opening an excessive number of editors. When this option is off, the debugger will open a new editor for each source file that needs to be displayed.
```

Allegedly `org.eclipse.debug.internal.ui.sourcelookup.SourceLookupFacility.display(ISourceLookupResult, IWorkbenchPage)` honors this preference via: `org.eclipse.debug.internal.ui.sourcelookup.SourceLookupFacility.openEditor(ISourceLookupResult, IWorkbenchPage)`

danthe1st replied:
My implementation opens one editor per file even if the preference is enabled (I think it's the same for the lambda-based implementation). I'll take a look but I don't know whether I can reuse the existing logic (it might be necessary to reimplement the logic or make `SourceLookupFactory#display` return the `IEditorPart`).

danthe1st replied:
Done in 4a6b692dc9fde1965f50a7e288db576ad920adae using the new `openInEditor` method which uses `SourceLookupFactory#display`.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaStackFrameSourceDisplayAdapter.java:
```diff
@@ -0,0 +1,159 @@
+		IJavaElement je = JavaUI.getEditorInputJavaElement(input);
+		if (je != null) {
+			IEditorPart part = JavaUI.openInEditor(je);
+			if (part instanceof ClassFileEditor editor && editor.getDocumentProvider().getDocument(editor.getEditorInput()).getLength() == 0) {
```
Can you add a comment that explains what this check does?

danthe1st replied:
Done in 4a6b692dc9fde1965f50a7e288db576ad920adae.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/sourcelookup/JavaStackFrameSourceDisplayAdapter.java:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	private boolean classFileUnhighlighterRegistered = false;
```
I believe this is checked in the UI thread, but set from a non-UI thread? I don't think debug events are propagated in the UI thread at least and display source sounds like UI thread code.

Can you debug and check? Best to use `AtomicBoolean` if that is the case. Or maybe `volatile` is enough.

I'm also not sure what happens if a debug event unregisters the listener concurrently to the code checking it it should be registered.

I'm also not sure about `displaySource` just running the UI code, e.g. `org.eclipse.debug.internal.ui.elements.adapters.StackFrameSourceDisplayAdapter.displaySource(Object, IWorkbenchPage, boolean)` schedules a job to do that...

danthe1st replied:
I think it is called in a UI thread, at least in my testing.
<img width="814" height="607" alt="image" src="https://github.com/user-attachments/assets/edcb6579-d39f-4ec4-b22a-a96331033f61" />

trancexpress replied:
I went through `StackFrameSourceDisplayAdapter` again, it looks like the job is started to not run the source look up in the UI thread.

So likely this is a problem already with: https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/812

danthe1st replied:
I switched to using `synchronized` in b07279bcc503b3dc180b464683e33c89cbe81ab6.

The reason for using `synchronized` over `volatile`/atomic variables is that there are multiple variables involved and if something tries to call `displaySource()` concurrently for whatever reason (I don't think this should happen under normal circumstances but I don't know for sure), it should still ensure that these variables are updated together and there's no hard to debug/reproduce in-between state.

trancexpress commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/sourcelookup/ClassFileEditorHighlightingTest.java:
```diff
@@ -0,0 +1,235 @@
+			ClassFileEditor editor = (ClassFileEditor) EditorUtility.openInEditor(javaProject.getProject().getFile("bin/" + CLASS_NAME + ".class"));
+			IClassFileEditorInput editorInput = (IClassFileEditorInput) editor.getEditorInput();
+			IClassFile classFile = editorInput.getClassFile();
+			thread.getTopStackFrame().getLaunch().setSourceLocator(stackFrame -> classFile);
```
Why is this necessary?

danthe1st replied:
This makes sure it actually uses the `ClassFileEditor` for displaying the source.
This is necessary because the source file is present in the project and otherwise not show the bytecode.

trancexpress commented:
I think https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/812 went in the wrong direction and this PR follows with more code.

With the addition of `LambdaStackFrameSourceDisplayAdapter`, every time I do debug steps inside lambda code, there is a source lookup call in the UI thread: `SourceLookupFacility.lookup()`

With this PR adding on top, a source lookup is done on every debug step - in the UI thread.

Normally the chain is: `SourceLookupFacility.displaySource()` -> `SourceLookupJob` (non-UI) -> `SourceDisplayJob` (UI) -> `SourceLookupFacility.display()` (opens editors, selects correct code)

The goal here must have been to not block the UI thread with source lookups - highlighting the current stack frame in the editor is secondary to smooth UI experience. I can imagine the source lookup causing noticeable mini-freezes in big workspaces (the effect would be tickets like https://github.com/eclipse-platform/eclipse.platform/issues/2615).

Since the source lookup and display chain is skipped, custom code is added in `LambdaStackFrameSourceDisplayAdapter` and was added in this PR. Such code then needs to duplicate preference handling and such (missing in `LambdaStackFrameSourceDisplayAdapter` but added in this PR).

@SougandhS IMO we need to clean this up, this PR is a good opportunity to do so - since even more intricate code must run once the editor is opened. I need to think on how we want to operate on the editor (given the stack frame), once its opened in `SourceLookupFacility.display()`... it would be great if you can suggest ideas. Maybe we are in the wrong adapter code, I'm not sure. Maybe we need something new. Maybe this: `org.eclipse.debug.ui.IDebugEditorPresentation.addAnnotations(IEditorPart, IStackFrame)`

```
	/**
	 * Positions and adds annotations to the given editor for the specified
	 * stack frame and returns whether any annotations were added. When
	 * <code>true</code> is returned, a call will be made to remove annotations
	 * when the source selection is cleared for the stack frame. When
	 * <code>false</code> is returned, the debugger will position and add
	 * standard annotations to the editor, and a corresponding call to remove
	 * annotations will not be made. This method is called when the debugger is
	 * has opened an editor to display source for the given stack frame.
	 *
	 * @param editorPart the editor the debugger has opened
	 * @param frame the stack frame for which the debugger is displaying
	 *  source
	 * @return <code>true</code> if annotations were added to the given editor part <code>false</code> otherwise
	 */
	boolean addAnnotations(IEditorPart editorPart, IStackFrame frame);
```

Snippet where source lookups are called on every step inside lambda:

```
	public static void main(String[] args) {
		int n = 3;
		int m = 20;
		Runnable r = () -> {
		for (int i = 0; i < n; ++i) {
			String l = "0".repeat(m);
			System.out.println(l);
		}
		};
		r.run();
	}
```

Once within the `Runnable.run()` body the lambda adapter runs its lookups.

SougandhS commented:
> I think #812 went in the wrong direction and this PR follows with more code.

> @SougandhS IMO we need to clean this up, this PR is a good opportunity to do so - since even more intricate code must run once the editor is opened. I need to think on how we want to operate on the editor (given the stack frame), once its opened in `SourceLookupFacility.display()`... it would be great if you can suggest ideas. Maybe we are in the wrong adapter code, I'm not sure. Maybe we need something new. Maybe this: `org.eclipse.debug.ui.IDebugEditorPresentation.addAnnotations(IEditorPart, IStackFrame)`

Handled this in a new [PR](https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/933) 👍

danthe1st commented:
> Handled this in a new [PR](https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/933) 👍

I guess my PR should wait for that and I can update it once that's merged. Hence, I'm marking this as a draft for now.

trancexpress commented:
@danthe1st please hold off rebasing, I'll try to adjust the test and I'll rebase in the process.

danthe1st commented:
> please hold off rebasing, I'll try to adjust the test and I'll rebase in the process.

Sorry, I didn't see that until now/I only saw the notification of the other PR being merged

trancexpress commented:
I've adjusted two things in the PR:

1. The tests no longer open the editor manually, IMO we want to test what the user will be seeing. This slows the tests somewhat, also due to creating the test project for each test. `testDisplaySourceWithClassFileEditorHighlightsLine` also compiles with debug line information, since otherwise Eclipse doesn't allow to create the line breakpoint.
2. The editor unhighlight code now runs in a UI job - I'm not sure the code belongs in the debug event notification thread, so I've moved it. Listener registration and deregistration is also done in the UI thread so no need to synchronize it anymore. @danthe1st please test if the unhighlight and listener deregistration works as expected, if this is not covered by the tests.

When I run the tests with coverage, `JavaStackFrameEditorPresenter.removeAnnotations(IEditorPart, IThread)` is not covered. I'm not sure when this code runs, we'll have to check. I'm also not sure if this code needs to unregister the listener.

I've also adjusted the test to have to run only a small part of the code in the UI thread, so I've removed the UI debug suite.

danthe1st commented:
> I'm not sure when this code runs, we'll have to check. I'm also not sure if this code needs to unregister the listener.

When I tested it, it didn't seem to be executed at all.

trancexpress commented:
> When I tested it, it didn't seem to be executed at all.

I'll try to find where its called.

One more thing to check:

What happens if I close the class editor and another editor is opened when stepping?  Will the unhighlighting work? Or if debugging multiple programs. Or the same program with multiple class files.

For the listener probably with different workbench windows we will have a problem too. Likely we need a list of registered listeners / highlighted editors, or some other mechanism that doesn't rely on fields (e.g. a task queue for the clean-up job). For the editors we likely must check if the editor is still opened before removing highlights.

I'm also not 100% sure what the life-cycle of `org.eclipse.jdt.internal.debug.ui.JDIModelPresentation` is.

Probably this is another case where we should check what platform debug is doing to remove line highlighting when the debug launch is terminated.

And it will be good to have a test for terminating the debug launch and checking that the active class editor has no highlight.

danthe1st commented:
> What happens if I close the class editor and another editor is opened when stepping? Will the unhighlighting work? Or if debugging multiple programs. Or the same program with multiple class files.

When I tested with what I tried yesterday (a listener that was registered using the `IWorkbenchWindow` of the editor), unhighlighting worked. I can test it again later.

> For the listener probably with different workbench windows we will have a problem too.

I can check that as well.

trancexpress commented:
@danthe1st if you have time, can you first check where the Java editor line highlight is removed when terminating a Java debug launch? And what the call stacks are to get to there?

danthe1st commented:
If `addAnnotations` returns `true`, `removeAnnotations` will be called when the instruction pointer doesn't need to be shown any more (this includes the launch being terminated). Both methods are called in a UI thread. I will update my PR accordingly.

danthe1st commented:
I now used `removeAnnotations` for the unhighlighting and I added an assertion that ensures this is actually done after termination. However, I think I noticed an NPE in that class so I'll fix that as well.

trancexpress commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/sourcelookup/ClassFileEditorHighlightingTest.java:
```diff
@@ -0,0 +1,327 @@
+			expectHighlightedText("    17  return", TIMEOUT);
+		} finally {
+			terminateAndRemove(thread);
+			expectNothingHighlighted(TIMEOUT);
```
This doesn't work here, unfortunately. Since if it fails, we'll leak the project (and maybe the breakpoints) to other tests.

I think you can add explicit terminating to the `try` block and checking for no highlight, and then leave the `finally` block as is. The terminate in `finally` is just in case, it doesn't hurt to terminate before that, in the actual test.

danthe1st replied:
Done in 356c27c31c024c715b34835fb31a5a6621806bc5.

trancexpress commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JavaStackFrameEditorPresenter.java:
```diff
@@ -66,7 +78,10 @@ public boolean addAnnotations(IEditorPart editor, IStackFrame frame) {
-		// nothing to clean up
+		if (editorPart instanceof ClassFileEditor classFileEditor
+				&& classFileEditor.getDocumentProvider().getDocument(editorPart.getEditorInput()).getLength() == 0) {
+			classFileEditor.unhighlight();
```
Yes, this looks very good now!

trancexpress commented:
Just the test issue with checking for no highlighting in `finally`, then we can let Copilot take a final look and I think we can merge.

danthe1st commented:
I fixed the aforementioned NPE which was caused by `getDocumentProvider()` returning `null` in `removeAnnotations()` if the editor was disposed (e.g. if you close the editor and then resume) and addressed your comment in 356c27c31c024c715b34835fb31a5a6621806bc5.

trancexpress commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/sourcelookup/ClassFileEditorHighlightingTest.java:
```diff
@@ -0,0 +1,361 @@
+			StyleRange[] styleRanges = noSourceTextWidget.getStyleRanges();
+			StyledTextContent content = noSourceTextWidget.getContent();
+			for (int i = 0; i < styleRanges.length; ++i) {
+				String highlightedText = content.getTextRange(styleRanges[0].start, styleRanges[0].length);
```
> styleRanges[0] instead of styleRanges[i]

Oops.

danthe1st replied:
Replaced with for-each in 987dba925f2a23a66244364fe179ea6417ef8af7.

danthe1st commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIStackFrame.java:
```diff
@@ -1756,6 +1756,11 @@ private boolean isDifferentLambdaContext(IStackFrame frame) throws DebugExceptio
 
+	public long getCodeIndex() {
+		synchronized (fThread) {
+			return fLocation.codeIndex();
```
Is there really a point from doing that for `codeIndex()`? From what I know, this is always just a field and cannot be unavailable (in contrast to e.g. line numbers).

trancexpress replied:
I think we don't need the try-catch with the current state.

danthe1st commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/sourcelookup/ClassFileEditorHighlightingTest.java:
```diff
@@ -0,0 +1,361 @@
+		DiagnosticCollector<JavaFileObject> collector = new DiagnosticCollector<>();
+		CompilationTask task = compiler.getTask(null, fileManager, collector, compilerOptions, null, List.of(fileObject));
+		Boolean result = task.call();
+		assertTrue(String.valueOf(collector.getDiagnostics()), result);
```
Done in 987dba925f2a23a66244364fe179ea6417ef8af7

danthe1st commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JavaStackFrameEditorPresenter.java:
```diff
@@ -66,7 +83,16 @@ public boolean addAnnotations(IEditorPart editor, IStackFrame frame) {
+		}
+		if (documentProvider.getDocument(editorPart.getEditorInput()).getLength() == 0) {
+			classFileEditor.unhighlight();
+		}
```
Done in 987dba925f2a23a66244364fe179ea6417ef8af7.

danthe1st commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JavaStackFrameEditorPresenter.java:
```diff
@@ -58,6 +64,17 @@ public boolean addAnnotations(IEditorPart editor, IStackFrame frame) {
+					methodName = methodName.substring(methodName.lastIndexOf('.') + 1);
+				}
+
+				classFileEditor.highlightInstruction(methodName, jdiFrame.getSignature(), jdiFrame.getCodeIndex());
```
Done in 987dba925f2a23a66244364fe179ea6417ef8af7.

danthe1st commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/sourcelookup/ClassFileEditorHighlightingTest.java:
```diff
@@ -0,0 +1,361 @@
+					if (status.matches(IStatus.ERROR)) {
+						errors.add(status);
+					}
+				}
+			});
+			resumeAndExit(thread);
```
Done in 070e8dcdc57ceb77db6e38fda93d545f27b8cb37.

trancexpress commented:
Thank you @danthe1st !

SougandhS commented:
@danthe1st please add an N&N entry for this feature 👍

danthe1st commented:
> please add an N&N entry for this feature 👍

I created https://github.com/eclipse-platform/www.eclipse.org-eclipse/pull/524.
